src/services/fs/backend.rs

```diff
@@ -126,6 +126,13 @@ pub struct Backend {
     atomic_write_dir: Option<String>,
 }
 
+#[inline]
+fn tmp_file_of(path: &str) -> String {
```

Comment:
How about calling this in `ensure_write_abs_path` and `blocking_ensure_write_abs_path` directly?

Reply:
`ensure_write_abs_path` and `blocking_ensure_write_abs_path` are also called with the target path, so we can't do this in them.